### PR TITLE
Add memory authentication tree software support

### PR DESCRIPTION
**Pull request by tianyouw**
https://github.com/tianyouw/minRISC/pull/3

### CODE REVIEW

File: projects/briey/libs/linker.ld [hunk 1]
```diff
@@ -28,7 +28,8 @@ MEMORY {
   sdram : ORIGIN = 0x40000000, LENGTH = 128M
 }
 _stack_size = DEFINED(_stack_size) ? _stack_size : 2K;
-_heap_size = LENGTH(sdram) - _stack_size - SIZEOF(.noinit); /* Fill rest of SDRAM with heap */
+_heap_size = LENGTH(sdram) - _stack_size - _memTree_size - SIZEOF(.noinit); /* Fill rest of SDRAM with heap */
+_memTree_size = LENGTH(sdram) / 2 ; /* Size of mem. auth. tree */
```
Comment (Jiangyi) [inline projects/briey/libs/linker.ld:31]: Shouldn't we declare _memTree_size before _heap_size?

File: projects/briey/memtest/src/main.c [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-#define PATTERN       (0b10101010u)
+#define PATTERN       (0x55555555U) // Repeating 0b0101...
 
 #define UART_DATA_LEN (8U)
 #define UART_BAUD     (115200U)
 
 extern void flushDataCache(); // From /vga/src/crt.S
-extern int _heap_start, _heap_end;
-volatile uint8_t state = 0;
+extern char _heap_start, _heap_end, _memTree_start, _memTree_end;
```
Comment (Jiangyi) [inline projects/briey/memtest/src/main.c:13]: Why the change to char?

Comment (tianyouw) [reply projects/briey/memtest/src/main.c:13]: To follow example at the bottom of: https://sourceware.org/binutils/docs/ld/Source-Code-Reference.html#Source-Code-Reference

**Review by Jiangyi — APPROVED** [review]

File: projects/briey/libs/stdlib.c [hunk 3]
```diff
@@ -43,22 +43,40 @@ static void printf_s(char *p)
...
+    uint32_t myVal = (uint32_t) val;
+    uint32_t len = 8;
+
+    while (len--) {
+        hexChar = myVal & 0x0F; // Mask least sig. hex digit
+        *(p++) = hexChar > 9 ? (hexChar + 'A' - 10) : (hexChar + '0');
+        myVal = myVal / 16;
+    }
+    while (p != buffer)
+        printf("%c", *(--p));
```
Comment (Denton-L) [inline projects/briey/libs/stdlib.c]: we shouldn't recursively call printf, use printf_c or whatever instead

Comment (tianyouw) [reply projects/briey/libs/stdlib.c]: Good catch, accidentally left that in after debugging.

File: projects/briey/memtest/src/main.c [hunk 4]
```diff
@@ -1,19 +1,20 @@
 #include <stdint.h>
 #include <stdlib.h>
+#include <stdio.h>
 #include <briey.h>
 
-#define PATTERN       (0b10101010u)
+#define PATTERN       (0x55555555U) // Repeating 0b0101...
```
Comment (Jiangyi) [inline projects/briey/memtest/src/main.c:7]: Could we go back to writing 1 byte at a time please? That would be a more comprehensive test in making sure cache readback isn't broken by the changes

**Review by Jiangyi — CHANGES_REQUESTED** [review]

File: projects/briey/memtest/src/main.c (hunk 4 above)
Comment (tianyouw) [reply projects/briey/memtest/src/main.c:7]: From offline discussion: leaving it as word transfers, adding todo for adding compile flag to test byte transfers.

**Review by Denton-L — APPROVED** [review]
